feat: gitlab skip source, target and labels

paolomainardi opened the pull request:
### **User description**
Related to #1190

This is a first attempt to support on the gitlab server the ability to skip automatic commands for source, target branches and labels.


___

### **PR Type**
Enhancement


___

### **Description**
- Implemented new features to ignore PRs/MRs based on title, labels, source branches, and target branches for both GitHub and GitLab
- Added configuration options in `configuration.toml` to support these new ignore features
- Updated GitHub and GitLab webhook handlers to check for ignore conditions before processing PRs/MRs
- Added comprehensive documentation explaining how to use the new ignore features
- Moved some configuration settings to a common section for better organization
- Standardized the implementation across GitHub and GitLab platforms



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>github_app.py</strong><dd><code>Enhance GitHub PR filtering capabilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/servers/github_app.py

<li>Moved configuration for ignoring PR titles to a common config section<br> <li> Added logic to ignore PRs with specific labels, source branches, or <br>target branches<br> <li> Implemented checks for these new ignore conditions<br>


</details>


  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1192/files#diff-3fb3f174152732d1b69953c8bd81b8a0a30f0e42100dc626d932da8371851608">+24/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gitlab_webhook.py</strong><dd><code>Implement GitLab MR filtering features</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/servers/gitlab_webhook.py

<li>Added logic to ignore merge requests based on title, labels, source <br>branches, and target branches<br> <li> Implemented checks for these new ignore conditions<br> <li> Imported 're' module for regex operations<br>


</details>


  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1192/files#diff-c3ea1ba460930c6e6e6a2c36a7dc729d79730a96e872d40ffb246f80bc7c0880">+32/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>additional_configurations.md</strong><dd><code>Document new PR/MR ignore configurations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/docs/usage-guide/additional_configurations.md

<li>Added documentation for new PR/MR ignoring features<br> <li> Provided examples of how to configure ignore settings in the <br>configuration.toml file<br>


</details>


  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1192/files#diff-9290a3ad9a86690b31f0450b77acd37ef1914b41fabc8a08682d4da433a77f90">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>configuration.toml</strong><dd><code>Update configuration with new ignore options</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/settings/configuration.toml

<li>Added new configuration options for ignoring PRs/MRs based on title, <br>labels, source branches, and target branches<br> <li> Moved some GitHub-specific settings to a common config section<br> <li> Added similar ignore settings for GitLab<br>


</details>


  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1192/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78">+18/-2</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

mrT23 commented:
This PR in its current state is far from being considered to merge, from several reasons:

1) The need is not strong enough. I have not encountered requests from users for this specific ability (ignore _branches_), and I don't there is a real need for that
2) Its gitlab only. New abilities, if possible, should be for all providers
3) **undocumented**.  Adding new params to configuration Should always be accompanied by proper documentation, and cannot be reviewed without it.

What can be done:
- Currently, there is some semi-maintained parameter, only for github app: `GITHUB_APP.IGNORE_PR_TITLE`
https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/servers/github_app.py#L141
- This parameter in practice is quite useful (for example, ignoring PRs that contain `[bump]`, or other keywords that symbolize something without real content_. 
- I would consider a PR for upgrading `IGNORE_PR_TITLE`:
    - moving it to the config
    - implementing it also for gitlab and bitbucket
    - documenting it

paolomainardi commented:
Thanks for your review, but I was pretty sure of the answer.

Let me reply to each point:

1. I think that excluding branches is useful in many cases, such as merging from one branch to another or releasing to the main. Labels can also be used to exclude MR opened by bots, such as RenovateBot. I don't think to be the only one who needs this, they are common scenarios.

We are discussing three parameters and two lines of code to handle them, it would be more or less the same for the other providers.

2. and 3. This is right, but the intention was to start discussing it first and then spend some more time on it. I guess I made the right choice to wait, as it has already been deprecated.

I'll look into the title stuff, but not in this PR, as they are totally unrelated.

I instead noticed that the GitHub provider has the ability to automatically skip "bot" PRs. This isn't something that the Gitlab server handles. It would be useful to add it, too, and maybe it can partially reduce the need for this MR, just partially.

paolomainardi commented:
Ok, I found where Github handles the bot users: https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/servers/github_app.py#L263 (anyway this is `dead configuration`), we cannot do the same on Gitlab has the concept of `bot` is not represented in any form in the webhook events, it should be just guessed: https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#merge-request-events

So this means not having a way on Gitlab to exclude RenovateBot automatic MR, the only way would be to exclude the the source branch `dependencies/*` or the label `dependencies` always added.

Could this be a good justification to implement it?

paolomainardi commented:
/describe

paolomainardi commented:
/review

paolomainardi commented:
/improve

paolomainardi commented:
@mrT23, just in case, I made some improvements:

1. Implemented the same logic for Github to skip source, target (list of regexp), labels
2. Added to Gitlab the title skip feature

mrT23 commented:
@paolomainardi 
ok, this is an improvement. To be able to merge the PR, make the following changes:

1) remove the 'label'. it's not useful, and also bug-prone (some users define labels only after the PR is opened, and its a race condition)
2) move the 3 new parameters to a unified [config] section, since they are the same for all platforms (I will later add them also to bitbucket)
3) add some line documenting the params in:
https://pr-agent-docs.codium.ai/usage-guide/additional_configurations/
(https://github.com/Codium-ai/pr-agent/blob/main/docs/docs/usage-guide/additional_configurations.md)

paolomainardi commented:
Thanks a lot, @mrT23. I just pushed the requested changes but labels, which I pray you keep as they are very useful. Automated systems like RenovateBot may open PRs with attached labels that can be easier to identify instead of using regexp titles.

Just as a reference, this is how I am currently using the feature:

```
config__ignore_pr_source_branches="['develop', 'main', 'master', 'stage']"
config__ignore_pr_labels="['dependencies']"
config__ignore_pr_title="['\\[AUTO\\]', '\\[SKIP-PR-AGENT\\]']"
```

paolomainardi commented:
/describe

paolomainardi commented:
/review

paolomainardi commented:
/improve

mrT23 commented on pr_agent/settings/configuration.toml:
```diff
@@ -230,6 +238,14 @@ push_commands = [
     "/describe",
     "/review --pr_reviewer.num_code_suggestions=0",
 ]
+# a list of regular expressions to match against the PR title to ignore the PR agent
+ignore_mr_title = []
+# target branches to ignore from PR agent when an MR is created
+ignore_mr_target_branches = []
+# source branches to ignore from PR agent when an MR is created
+ignore_mr_source_branches = []
+# labels to ignore from PR agent when an MR is created
+ignore_mr_labels = []
```
@paolomainardi remove this, via the [config]
I want to have unified params. they are easier to maintain, track an use

paolomainardi replied:
Ooops, it's a leftover, I'll remove it.

mrT23 commented:
ok, i will merge, but i will later remove 'labels' from the instructions.
Its just a plain mistake - all the bots will **compete** against PR-agent, and their feedback will probably won't be available at the beginning of the PR, when PR-Agent pulls the PR. So problems are almost guranteed.

paolomainardi commented:
Thanks for merging @mrT23.

Still don't understand what's the point with labels, the idea behind this is that a PR is opened with certain labels, it will be skipped, exactly as with the title.

Why it should be confusing or misleading ? Genuinely curious, maybe I am missing something.
